alarm_base/src/frame.rs

```diff
@@ -1 +1,29 @@
-//! Base types for page frame allocators.
\ No newline at end of file
+//! Base types for page frame allocators.
+use alloc::allocator::AllocErr;
+
+/// An allocator that provides page frames.
+pub unsafe trait Allocator {
+
+    /// Architecture-dependent size of a physical page.
+    const FRAME_SIZE: usize;
+
+    /// Type representing frames provided by this allocator.
+    ///
+    /// A `Frame` must either be a pointer to a contiguous block of `FRAME_SIZE`
+    /// bytes, or be a handle that may be converted into such a pointer.
+    type Frame;
+
+    /// Returns a new `Frame`.
+    unsafe fn alloc(&mut self) -> Result<Self::Frame, AllocErr>;
+
+    /// Deallocate a `Frame`.
+    ///
+    /// # Unsafety
+    /// This function is unsafe because undefined behaviour may result if the
+    /// given `frame` was not originally allocated by this `Allocator`.
+    unsafe fn deallocate(&mut self, frame: Self::Frame);
```

Comment:
Should `deallocate` not return whether it succeeded since it can go wrong?
Also, maybe it should be called `dealloc` to be consistent with `alloc`

Reply:
> Should `deallocate` not return whether it succeeded since it can go wrong?

Hmm. Normally, I'd say "yes, definitely", but the tricky bit here is that there isn't really an easy way for the allocator to *know* when a given frame shouldn't be deallocated. The errors that could occur here are when a frame was allocated by a different allocator (which we might be able to detect if we know what memory range our frames come from...) or if it's still in use somewhere, which we have no real way to know about at all. Which is why this entire API is massively `unsafe`...

> Also, maybe it should be called `dealloc` to be consistent with `alloc`
Yeah, good call 👍